psef/models/auto_test_step.py

```
@@ -0,0 +1,871 @@
+        >>> res
+        {'steps': [{'achieved_points': 5}]}
+        """
+        log_dict = t.cast(t.Dict, log if isinstance(log, dict) else {})
+        steps = log_dict.get('steps', [])
+        steps = [{'achieved_points': s.get('achieved_points')} for s in steps]
+        return {'steps': steps}
+
+
+@_register
+class _RunProgram(AutoTestStepBase):
+    __mapper_args__ = {
+        'polymorphic_identity': 'run_program',
+    }
+
+    def validate_data(self, data: JSONType) -> None:
+        with get_from_map_transaction(
+            ensure_json_dict(data), ensure_empty=True
+        ) as [get, _]:
+            program = get('program', str)
+        _ensure_program(program)
+
+    @staticmethod
+    def _execute(
+        container: 'auto_test_module.StartedContainer',
+        update_test_result: 'auto_test_module.UpdateResultFunction',
+        test_instructions: 'auto_test_module.StepInstructions',
+        total_points: float,
+    ) -> float:
+        data = test_instructions['data']
+        assert isinstance(data, dict)
+
+        res = 0.0
+
+        code, stdout, stderr, time_spend = container.run_student_command(
+            t.cast(str, data['program']),
+            test_instructions['command_time_limit'],
+        )
+
+        if code == 0:
+            state = AutoTestStepResultState.passed
+            res = test_instructions['weight']
+        else:
+            state = AutoTestStepResultState.failed
+
+        update_test_result(
+            state, {
+                'stdout': stdout,
+                'stderr': stderr,
+                'exit_code': code,
+                'time_spend': time_spend,
+            }
+        )
+
+        return res
+
+
+@_register
+class _CustomOutput(AutoTestStepBase):
+    __mapper_args__ = {
+        'polymorphic_identity': 'custom_output',
+    }
+
+    _MATCH_GROUP_NAME = 'amount_of_points'
+    _FLOAT_REGEX = r'(?P<{}>(?:-\s*)?(?:1(?:\.0*)?|0(?:\.\d*)?))'.format(
```

Since this now has a capture group, can it still be used multiple times in the regex when the output contains multiple floats?

---

Multiple `\f` in the regex compiles just fine at the moment, and the latest takes precedence. It is a bit strange, as it is very unclear what the behavior should be (if the regex is `\f \f` and the output is `0.5 0.75` what should we match), the default regex engine of Python for example raises an exception. However, while `\f \f` should maybe raise an exception, `(grade: \f|you got \f points)` should not in my opinion. Detecting if two regexes overlap is not really that easy though.

So I think it is fine as it is right now, as most users will probably not try it